drivers: watchdog: add support for agilex platform, reset and clock

pbalsundar opened the pull request:
- If MMU is enabled, the physical address has to be mapped to a virtual address at run-time. Added code changes to do the same using DEVICE_MMIO_* macros, which maps physical address of register space into a memory
- according to spec when the watchdog timer is expired and interrupt is raised, interrupt flag should not be cleared. Clearing interrupt flag will not assert a system reset. Fixed this issue by not clearing an interrupt flag
- added clock manager support to get clock frequency at runtime. If clock manager is present, then peripheral driver can get clock rate at run-time using clock manager.
- added support for watchdog reset through reset manager. If reset controller or reset manager driver is present to reset peripheral IPs, it can be used to reset watchdog timer to default state
- enabled/disabled watchdog timer at boot time based on config flag WDT_DISABLE_AT_BOOT
- pause while debug or pause while sleep is not configurable through application, so added a warning logs to the user
- Some boards may have reset lines connected either to SOC or reset line connected to CPU or reset line is not connected to any entity. But these resets are not configurable throught the application. So just logging a warning message to the user
- adding driver support for agilex and agile5 socfpga
- added agilex board support for watchdog sample application

pbalsundar committed:
samples: drivers: watchdog: add sample application support

add agilex and agilex5 board support in watchdog sample application

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar committed:
CODEOWNERS: update codeowner for watchdog

adding code owner for snps designware watchdog

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar commented:
Hi @nashif  @marcinszkudlinski  @softwarecki @carlocaione 
Could you please review and give necessary comments?

martinjaeger commented on drivers/watchdog/Kconfig.dw:
```diff
@@ -9,3 +9,21 @@ config WDT_DW
+if WDT_DW
+
+config WDT_DW_PAUSE_BY_DBG
+	bool "pause watchdog counter while debugging"
```
Please start with uppercase (for consistency).

pbalsundar replied:
updated

martinjaeger commented on drivers/watchdog/Kconfig.dw:
```diff
@@ -9,3 +9,21 @@ config WDT_DW
+	help
+	  pause mode for debugging
+
+config WDT_DW_RESET_MODE
+	int "reset none or SoC or cpu cores"
+	default 0
+	range 0 2
+	help
+	  select 0 for reset none
+	  select 1 for cpu core reset
+	  select 2 for SoC reset
```
Why not use a Kconfig `choice`?

pbalsundar committed:
dts: arm64: intel: intel_socfpga: Adding nodes for watchdog

add dts support for watchdog to accomodate watchdog driver bringup on
aglex and agilex5

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

carlocaione commented on drivers/watchdog/wdt_dw_common.c:
```diff
@@ -19,10 +19,12 @@ LOG_MODULE_REGISTER(wdt_dw_common, CONFIG_WDT_LOG_LEVEL);
 
 int dw_wdt_check_options(const uint8_t options)
 {
+#if !CONFIG_WDT_DW_PAUSE_BY_DBG
```
you mean `#ifndef`?

pbalsundar replied:
updated

carlocaione commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -15,14 +15,38 @@
 #include "wdt_dw_common.h"
 
+/* Check if reset property is defined */
+#define WDT_DW_IS_INST_RESET_EN(inst)	DT_NODE_HAS_PROP(DT_DRV_INST(inst), resets)
+#define WDT_DW_CHECK_RESET(inst)	WDT_DW_IS_INST_RESET_EN(inst) ||
+#define WDT_DW_RESET_SUPPORT		(DT_INST_FOREACH_STATUS_OKAY(WDT_DW_CHECK_RESET) 0)
```
I'm confused by this. You are checking if a reset line exists in one of the instances in the DT? What about the other instances that do not have it?

carlocaione commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -15,14 +15,38 @@
+#endif
+
+/* Check if clock manager is supported */
+#define WDT_DW_IS_INST_CLK_MGR_EN(inst)	DT_PHA_HAS_CELL(DT_DRV_INST(inst), clocks, clkid)
+#define WDT_DW_CHECK_CLK_MGR(inst)	WDT_DW_IS_INST_CLK_MGR_EN(inst) ||
+#define WDT_DW_CLK_MANAGER_SUPPORT	(DT_INST_FOREACH_STATUS_OKAY(WDT_DW_CHECK_CLK_MGR) 0)
```
This really is macromagic and macrobatics

povergoing replied:
I am just curious if it is necessary to use `#if xxx` "everywhere"? Do we really need to lower the footprint? Just too many macros, I think, some of them can be replaced by C code?

pbalsundar replied:
I have updated the changes these with macro(DT_ANY_INST_HAS_PROP_STATUS_OKAY) defined in the devicetree.h, which is similar implementation

pbalsundar commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -11,18 +11,30 @@
-#define WDT_CHECK_INTERRUPT_USED(inst)	WDT_IS_INST_IRQ_EN(inst) ||
-#define WDT_DW_INTERRUPT_SUPPORT	DT_INST_FOREACH_STATUS_OKAY(WDT_CHECK_INTERRUPT_USED) 0
+/* Check if reset property is defined */
+#define WDT_DW_RESET_SUPPORT		DT_ANY_INST_HAS_PROP_STATUS_OKAY(resets)
```
updated

pbalsundar commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -11,18 +11,30 @@
+#define WDT_DW_RESET_SUPPORT		DT_ANY_INST_HAS_PROP_STATUS_OKAY(resets)
+
+/* Check if clock manager is supported */
+#define WDT_DW_CLK_MANAGER_SUPPORT	DT_ANY_INST_HAS_PROP_STATUS_OKAY(clocks)
```
updated

pbalsundar commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -11,18 +11,30 @@
+/* Check if clock manager is supported */
+#define WDT_DW_CLK_MANAGER_SUPPORT	DT_ANY_INST_HAS_PROP_STATUS_OKAY(clocks)
+
+#define WDT_DW_INTERRUPT_SUPPORT	DT_ANY_INST_HAS_PROP_STATUS_OKAY(interrupts)
```
updated

carlocaione reviewed (dismissed):
I don't have any more comments but this driver is a mess.

pbalsundar committed:
drivers: watchdog: wdt_dw: resetting CPU or SoC is not configurable

specific to platform, watchdog reset line can be connected either to
CPU, SOC or none of the entity. These resets cannot be configured from the
application. So added a warning message when application configures this
option

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar committed:
drivers: watchdog: wdt_dw: add watchdog disable at boot and pause by debug

added support for watchdog enable/disable at boot

Pausing watchdog timer when CPU is halted by the debugger and
pausing watchdog timer when CPU is in sleep state is not
configurable through application, so added warning log with return success

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar commented:
@martinjaeger sorry for late response and changes, I have made separate commits
1. adding clock through manager support, reset support, using mmio address, one fix related to clearing interrupt flag
2. adding support for enable/disable watchdog at boot and watchdog pause while debugging
3. Added config flags for differentiating the different reset type supported by the board through watchdog

Could you please review the changes?


@softwarecki, @katsuster, @SgrrZhf could you also please review and give necessary comments :-)

martinjaeger commented:
Thanks @pbalsundar for splitting up some commits. However, I'm still not completely happy. Correct me if I'm wrong, but if we look at the first commit for instance, the mentioned points look like two independent changes, so they should go into separate commits:

```
drivers: watchdog: adding driver support and optional feature support
added support for watchdog reset through reset manager
added clock manager support to get clock frequency runtime
```

Also the commit headline should contain the driver you are touching. You are not adding support for the first watchdog driver. Instead you are changing an existing one and adding support for another SOC of a similar kind, if I understood correctly. And if I look at the code, I see lots of MMIO macros being used. Please describe in the commit message why this change is necessary in this context. Please help me understanding your changes and making review more easy (keep in mind that I don't know anything about this Intel platform).

pbalsundar committed:
drivers: watchdog: wdt_dw: support for resetting watchdog

add support for resetting watchdog IP through reset manager driver

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar committed:
drivers: watchdog: wdt_dw: get clock rate from clock manager

added support for getting clock rate at runtime from clock manager

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar committed:
drivers: watchdog: wdt_dw: use mmio address in RAM

mmu enabled platform needs mapping of physical address ranges to
the virtual address at runtime. So using DEVICE_MMIO_* helper macros to
map physical csr address space to RAM runtime if MMU is enabled

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar committed:
drivers: watchdog: wdt_dw: bug fix and modify conditional macro check

clearing interrupt flag will not assert the system reset as per IP spec,
so interrupt flag should not be cleared in isr

use #if macro check directly with DT_ANY_INST_HAS_PROP_STATUS_OKAY for
interrupt

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

martinjaeger commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -88,8 +100,13 @@ static int dw_wdt_install_timeout(const struct device *dev, const struct wdt_tim
+#else
 	return dw_wdt_calc_period((uint32_t)reg_base, dev_config->clk_freq, config,
 				  &dev_data->config);
+#endif
```
I find it confusing to get the clock frequency either as a hard-coded value from devicetree or via `clock_control`. Can we not just enforce using `clock_control` and get rid of all these `#ifdef`s?

pbalsundar replied:
@martinjaeger, it is possible that for few platform there is no option for clock control and for some platform clock rate can ge queried through clock controller. I feel these both the methods are required.

@softwarecki, I can do that, assign a variable with clock rate from dev_data or dev_config, but still we will have a #if check.

martinjaeger replied:
Why not implement clock control where it is missing?

pbalsundar replied:
@martinjaeger, I think I did not understand this comment fully. Do you mean to remove the dependecy of driver to get clock frequecy from the dtsi for all the platforms? Could you please elaborate on this?

martinjaeger replied:
Ideally, you would implement the clock_control driver for the missing platforms in a separate PR, the that merged and afterwards rebase this one.

Your above comment was very vague, that "it is possible that for few platform there is no option for clock control". How many and which platforms are there, actually?

As a compromise I suggest to use `dev_data->clk_freq` all the time (even if the frequency is determined from devicetree at compile-time). This would avoid the `#ifdef`s.

softwarecki replied:
@pbalsundar Yes, #ifdef will remain, but it will not contain the entire function call, which will be more readable.

pbalsundar replied:
@martinjaeger @softwarecki I have added the clk_freq to the dev_data instead of dev_cfg structure.

martinjaeger commented on drivers/watchdog/Kconfig.dw:
```diff
@@ -9,3 +9,12 @@ config WDT_DW
+
+if WDT_DW
+
+config WDT_DW_PAUSE_BY_DBG
```
I can't see any location in the driver where the pause by debug functionality is enabled depending on this config being enabled. Am I missing something?

pbalsundar replied:
@martinjaeger it is present in wdt_dw_common.c

@softwarecki True, software configuration doesn't have any effect with respect to configuring synopsis dw IP.

Even if the platform design is such that it is configured to pause the watchdog, user will assume that he needs to configure that as a part of wdt_setup function, because it is defined in watchdog.h(user interface file).

For example user uses sample application and lets say hardware line pause by debug signal is connected to watchdog(which is true in Agilex5 platform). Sample application calls wdt_setup() function with the WDT_OPT_PAUSE_HALTED_BY_DBG. The wdt_dw driver will respond that the WDT_OPT_PAUSE_HALTED_BY_DBG is not supported even though user has sent a correct flag. But this is not true in case of agilex and agilex5 platform
This is the reason I introduced the flag. Let me know your view on this.

softwarecki replied:
In that case, it may be enough to modify the error message to inform user that it is not configurable by the software or simply ignore that flag. Personally, I wouldn't add a fake Kconfig option to bypass flag checking.

pbalsundar replied:
removed this option as this is not configurable through application and added log message to inform the user about the same if user tries to configure this option

martinjaeger commented on samples/drivers/watchdog/boards/intel_socfpga_agilex5_socdk.overlay:
Could all these nodes not go into the board dts file or even SOC dtsi files, leaving only the `status = "okay"` in the overlay?

pbalsundar replied:
we kept the interrupt configuration as a part of sample application overlay file, because we wanted to give the flexibility to the user that when the watchdog timer expires user wants to reset the system immediately(without configuring an ISR) or take some actions before the triggering system reset in ISR

pbalsundar commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -17,16 +17,12 @@
 	/* MMIO mapping information for watchdog register base address */
 	DEVICE_MMIO_RAM;
 	uint32_t config;
-#if WDT_DW_INTERRUPT_SUPPORT
```
Did that way initially, got a comment on this to not hide any parameter with macros.
https://github.com/zephyrproject-rtos/zephyr/pull/60393#discussion_r1277398077

pbalsundar replied:
similar reference for clocks property
https://github.com/zephyrproject-rtos/zephyr/pull/60393#discussion_r1283333773

interrupts:
https://github.com/zephyrproject-rtos/zephyr/pull/60393#discussion_r1283333888

pbalsundar commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -133,16 +133,18 @@ static int dw_wdt_init(const struct device *dev)
 
-		dw_wdt_clear_interrupt((uint32_t)reg_base);
-
+		/*
```
I would like to keep this as it is

pbalsundar commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -59,7 +71,7 @@ static int dw_wdt_setup(const struct device *dev, uint8_t options)
 static int dw_wdt_install_timeout(const struct device *dev, const struct wdt_timeout_cfg *config)
 {
-#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(interrupts)
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(interrupts) || !DT_ANY_INST_HAS_PROP_STATUS_OKAY(clocks)
```
Agree that it is getting complex, I will use __maybe_unused gcc attributes to fix this

pbalsundar commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -21,6 +22,10 @@ LOG_MODULE_REGISTER(wdt_dw, CONFIG_WDT_LOG_LEVEL);
 struct dw_wdt_dev_data {
 	/* MMIO mapping information for watchdog register base address */
 	DEVICE_MMIO_RAM;
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(clocks)
```
This needs to be #if DT_ANY_INST_HAS_PROP_STATUS_OKAY(clocks). I can do #if not defined to save the memory in case of clk_freq in dw_wdt_dev_config structure

softwarecki replied:
Indeed, my mistake. I confused the data and config structures.

pbalsundar commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -123,9 +128,42 @@ static int dw_wdt_feed(const struct device *dev, int channel_id)
 	return 0;
 }
 
+int dw_wdt_disable_wrapper(const struct device *dev)
```
I though initially of doing this change. But if I do this change your wdt_intel_adsp will be affected as it won't be able to find the definition. Shall I go forward and add definition in wdt_intel_adsp.c as return -ENOTSUP?

softwarecki replied:
This is what I meant to move `dw_wdt_disable` function to wdt_intel_adsp.

pbalsundar replied:
I have updated this

pbalsundar commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -96,10 +96,22 @@ static int dw_wdt_install_timeout(const struct device *dev, const struct wdt_tim
+		LOG_ERR("only SoC reset supported");
+		return -ENOTSUP;
+	}
+#endif
```
I agree there is no IP configuration that has to be done by the driver for resetting and it is completely hardware platform design. In agilex5 platform, board supports a system reset through watchdog expiry. Any technical documentation of the board will say that this supports SoC reset or CPU or none of resets based on the board design.

The user may pass the `struct wdt_timeout_cfg` to `wdt_install_timeout(`) function(for eg sample application passes WDT_FLAG_SOC_RESET_SOC). User can pass the `WDT_FLAG_RESET_SOC `flag and the watchdog driver will fail to install the timeout by responding that this is not supported.

I can think of two solution
1. don't check this flag when `wdt_install_timeout()` is called or 
2. use these three different config flags, so that driver will respond to the user properly as he knows what to configure while installing timeout

softwarecki replied:
Personally, I would ignore the flag or tell the user to not use it because it is not configurable by the software.

pbalsundar replied:
Ignored this with adding warning logs to let the user know that this option is not cofigurable through application

martinjaeger commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -11,55 +11,77 @@
+	DEVICE_MMIO_RAM;
+	/* Clock frequency */
+	uint32_t clk_freq;
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(clocks)
+	/* Clock controller dev instance */
+	const struct device *clk_dev;
+	/* Identifier for timer to get clock freq from clk manager */
+	clock_control_subsys_t clkid;
+#endif
```
The clock controller information (if existing) is still constant and can stay in the config struct, right?

pbalsundar replied:
yeah, clock controller information is constant always. Just to make it look simple, I created it in the data struct. I can make these changes and add it in config struct

pbalsundar commented:
Hi @martinjaeger @carlocaione @softwarecki @povergoing 
I have made the requested changes. Could you please review further and give necessary comments?

pbalsundar commented:
Hi @galak 
I think the twister is failed due to unknwn reason. I couldn't find the log information for the failure
Failed: **Run tests with twister / twister-build (4) (pull_request_target)** 

Do we need to re-run this again? I couldn't see any option to re-run it. Could you please help me in re-running the failed twister build?

kartben commented:
@pbalsundar I re-triggered the CI job

pbalsundar commented:
@kartben Thanks for re-triggering it 👍

martinjaeger commented on drivers/watchdog/wdt_dw.c:
```diff
@@ -11,55 +11,77 @@
 static int dw_wdt_setup(const struct device *dev, uint8_t options)
 {
-	const struct dw_wdt_dev_cfg *const dev_config = dev->config;
+	int ret = 0;
 	struct dw_wdt_dev_data *const dev_data = dev->data;
+	uintptr_t reg_base = DEVICE_MMIO_GET(dev);
```
Unnecessary assignment. And please place the `ret` declaration at the end (more common in Zephyr).
```suggestion
	struct dw_wdt_dev_data *const dev_data = dev->data;
	uintptr_t reg_base = DEVICE_MMIO_GET(dev);
	int ret;
```

pbalsundar replied:
removed the initial value assignment which was not needed

martinjaeger commented:
I assume you have run all related tests on your hardware and they are all passing.

pbalsundar commented:
@martinjaeger Yes the driver is tested on hardware. Thanks for the approval
@softwarecki Thanks for the approval

pbalsundar commented:
Hi @martinjaeger  @katsuster
There are two approvals for this PR, can we merge it with two approvals?
